fix(git_status): show working tree status if index status is present

EliaGeretto opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!--- To help with semantic versioning the PR title should start with one of the conventional commit types. -->
<!--- The conventional commit types for Semantic PR are: feat, fix, docs, style, refactor, perf, test, build, ci, chore, revert -->

#### Description
This PR modifies the parser for the output of the `git` command in the `git_status` module. It ensures that, even when an index status is present or the file has been copied/renamed, the working tree status is correctly parsed and displayed in the prompt. This does not apply to merge conflict situations, in which only the merge conflict indicator is shown, as before.

#### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
This PR closes #2937.

In addition, files that were both renamed and modified were only shown as renamed, but not as modified.

#### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
- [ ] I have tested using **MacOS**
- [X] I have tested using **Linux**
- [ ] I have tested using **Windows**

#### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I have updated the documentation accordingly.
- [X] I have updated the tests accordingly.

andytom commented on src/modules/git_status.rs:
```diff
@@ -230,43 +230,46 @@ struct RepoStatus {
-        self.modified += Self::is_modified(s) as usize;
-        self.staged += Self::is_staged(s) as usize;
-        self.untracked += Self::is_untracked(s) as usize;
+        match s.chars().next().unwrap() {
+            '1' => self.parse_normal_status(&s[2..4]),
+            '2' => {
+                self.renamed += 1;
+                self.parse_normal_status(&s[2..4])
+            }
+            'u' => self.conflicted += 1,
+            '?' => self.untracked += 1,
+            _ => panic!("git_status: unknown line type"),
```
Rather than unwrapping here it would be better to include it in the match, something like (I am on my phone so this might not be completely correct) 
```suggestion
        match s.chars().next() {
           Some('1') => self.parse_normal_status(&s[2..4]),
           Some('2') => {
                self.renamed += 1;
                self.parse_normal_status(&s[2..4])
            }
            Some('u') => self.conflicted += 1,
            Some('?') => self.untracked += 1,
            _ => panic!("git_status: unknown line type"),
```

andytom replied:
Rather than panicking here it would be better to log an error.

andytom reviewed (changes requested):
Thank you for your contribution @EliaGeretto, I have left some suggestions to make the code a bit more robust but other than that this looks OK so far.

EliaGeretto commented:
The logic behind panicking was that an error in those two cases indicates a programming error somewhere, because the `git status --porcelain=2` format is defined and guaranteed not to change. Anyway, I have updated the commit to avoid panicking on a parsing error and, instead, write to the log. I also realized that I was not handling the `!` line type, which should not show, but is still valid, so I added it as a case to the `match`.

andytom commented on src/modules/git_status.rs:
```diff
@@ -230,43 +230,51 @@ struct RepoStatus {
-        self.modified += Self::is_modified(s) as usize;
-        self.staged += Self::is_staged(s) as usize;
-        self.untracked += Self::is_untracked(s) as usize;
+        if let Some(line_type) = s.chars().next() {
+            match line_type {
+                '1' => self.parse_normal_status(&s[2..4]),
+                '2' => {
+                    self.renamed += 1;
+                    self.parse_normal_status(&s[2..4])
+                }
+                'u' => self.conflicted += 1,
+                '?' => self.untracked += 1,
+                '!' => (),
+                _ => log::error!("Unknown line type in git status output"),
+            }
+        } else {
+            log::error!("Missing line type in git status output");
+        }
```
Not a blocker but I think the if let can be removed from here
```suggestion
           match s.chars().next() {
              Some('1') => self.parse_normal_status(&s[2..4]),
              Some('2') => {
                  self.renamed += 1;
                  self.parse_normal_status(&s[2..4])
              }
              Some('u') => self.conflicted += 1,
              Some('?') => self.untracked += 1,
              Some('!') => (),
              Some(_) => log::error!("Unknown line type in git status output"),
              _ => log::error!("Missing line type in git status output"),
             }
```

andytom reviewed (approved):
Have left a little suggestion but other than that I think this is ok

andytom commented:
> The logic behind panicking was that an error in those two cases indicates a programming error somewhere, because the `git status --porcelain=2` format is defined and guaranteed not to change. Anyway, I have updated the commit to avoid panicking on a parsing error and, instead, write to the log. I also realized that I was not handling the `!` line type, which should not show, but is still valid, so I added it as a case to the `match`. I also added a test case that verifies that ignored files are not shown.

In general we try to avoid panics whenever we can as that will cause starship to not output anything which means that the user doesn't get a prompt which is a very bad user experience, while the output of `git status --porcelain=2` should never change there is always the possibility that something else causes the output to get mangled (e.g a wrapper script around git) so we always try to be defensive in these sorts of situations.

EliaGeretto commented:
I have applied the modification requested in the comment, formatted and squashed. The PR should be ready to be merged.

andytom commented:
Thank you for your contribution @EliaGeretto.
